proposed/2023/NuGetAudit-without-nuget.org-as-packageSource.md

```diff
@@ -0,0 +1,217 @@
+Someone at Microsoft published [a whitepaper titled "3 ways to mitigate risk when using private package feeds"](https://azure.microsoft.com/mediahandler/files/resourcefiles/3-ways-to-mitigate-risk-using-private-package-feeds/3%20Ways%20to%20Mitigate%20Risk%20When%20Using%20Private%20Package%20Feeds%20-%20v1.0.pdf), so from customers points of view, it appears that Microsoft's official advice is to not use nuget.org directly, but instead use a single package source that uses nuget.org as an upstream source.
+
+While the design of NuGetAudit attempts to make it easy for upstreaming package sources to also upstream the `VulnerabilityInfo` resource, the NuGet team and customers are reliant of the feed implementor to implement the change.
+An upstreaming package source is one that either aggregates multiple package sources or caches packages from another package source.
+Upstreaming isn't a concept that exists in NuGet, so servers have to present to clients as if all data (packages, search results, vulnerability information) originates from itself.
+
+However, there are also cases where customers are not using upstreaming feeds, but instead curate approved 3rd party packages and manually upload them to an internal feed.
+There are also other scenarios where customers do not use nuget.org (directly) as a package source.
+
+It's possible for customers to use [Package Source Mapping](https://learn.microsoft.com/nuget/consume-packages/package-source-mapping) to add nuget.org as a package source, but configured in a way that NuGet will never restore packages from this package source.
+However, some developers might consider this an unacceptable risk and prefer another way to get a VDB that cannot accidentally be used to get packages.
+
+## Explanation
+
+### Functional explanation
+
+<!-- Explain the proposal as if it were already implemented and you're teaching it to another person. -->
+<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or  pseudocode to show how this proposal would look. -->
+
+#### auditSources in nuget.config
+
+*NuGet.Config* files have a new section `<auditSources>` where the URL(s) of any audit sources can be specified, which will be used only to download a VDB, and will not be used for downloading packages.
+Like `<packageSources>`, `<auditSources>` can contain zero, one, or more sources, and supports `<clear />`.
+When there is at least one source in `auditSources`, NuGet should no longer use `packageSources` to get a VDB.
+
+For example, consider a developer at Contoso using a company-internal source as a single package source, and using nuget.org as an audit source:
+
+```xml
+<configuration>
+  <packageSources>
+    <clear />
+    <add key="contoso" value="https://internal.corp.contoso.com/packages/nuget/index.json" >
```

Comment:
What this means is that `contoso` will *not* have its vulnerability resource checked, right?

If the `<auditSources>` tab were missing or did not contain a `<clear/>`, then both `contoso` and `nuget.org` would be checked for vulnerability resources?

Reply:
> What this means is that contoso will not have its vulnerability resource checked, right?

correct

> If the `<auditSources>` tab were missing or did not contain a <clear/>, then both contoso and nuget.org would be checked for vulnerability resources?

No. This is [the same interpretation as what Anand had](https://github.com/NuGet/Home/pull/12918#discussion_r1372141490).

The `<clear />` only means to remove any `auditSources` defined in nuget.org files, which is the same semantics as what `packageSources`, `packageSourceCredentials`, `packageSourceMapping`, and anything else in nuget.config that supports `<clear/>`.

I don't know NuGet.Configuration's implementation, so I don't know how easy or difficult it would be to have `auditSources` effectively auto-import all `packageSources`, and therefore use `<clear/>` to not only clear `auditSources` defined in different nuget.config files, but also clear `packageSources` that auto-flowed. But then we get into more complex scenarios like what if `c:\src\nuget.config` defines PkgSrc1 as a package source, and nuget.org as an audit source (without a clear), then `c:\src\sln1\nuget.config` sets up only `packageSources` with a clear and PkgSrc2, then a URL, but does not have an audit sources section.  What would the resulting audit sources be? PkgSrc1 and nuget.org because that's what was inherited from the parent directory? PkgSrc2 and nuget.org, because the "accumulated" package sources is PkgSrc2 only?

Also, would there be a desire to "clear inherited audit sources defined in a parent nuget.config, but I still want to include package sources as audit sources, in addition to adding an additional audit source?

This is why I really don't like using `packageSources` as a fallback for `auditSources`. It adds too much complexity and therefore confusion. I don't think defining the same URL twice, once for package sources and once for audit sources is a serious enough "problem" to warrant implicitly using the package sources, when customers do want to use the same source for both.

However, because we don't like to introduce breaking changes, this is the only reason I proposed continuing to use `packageSources` as a fallback when the `auditSources` list is empty. It would be less confusing for everyone is NuGetAudit never used `packageSources` to get audit information. But, it would be a breaking change.

This spec does not provide any way to stop using package sources as an audit source, without also explicitly defining at least 1 audit source. Because what that would mean is to disable NuGetAudit, which they can do via `<NuGetAudit>false</NuGetAudit>` in their project file or Directory.Build.props.

Reply:
I'm resolving this conversation as a duplicate of the one I linked. Please comment there to help me drive this spec to something that's sufficiently agreed upon so it can be merged.